Development

mosesogwo opened the pull request.

mosesogwo committed:
move enumerable.rb to lib folder

mosesogwo committed:
add enumerables spec file add format doc to rspec file

mosesogwo committed:
modified my_each to pass test

mosesogwo committed:
add tests for my_any, my_all, my_select

mosesogwo committed:
add tests for my_count, my_none

mosesogwo committed:
add tests for my_count, my_none

mosesogwo committed:
add tests for my_map

mosesogwo committed:
add tests for my_inject

mosesogwo committed:
add tests for #my_each_with_index

mosesogwo committed:
modified my_each_with_index

mosesogwo committed:
combined my_select tests

mosesogwo committed:
modified my any with extra example, empty array case

mosesogwo committed:
modified my_none test

mosesogwo committed:
complete my_map, stuck on inject, initial memo not resolved

mosesogwo committed:
update my_each method

mosesogwo committed:
update my_each_with_index method

mosesogwo committed:
update my_all method

mosesogwo committed:
update my_any test

mosesogwo committed:
update my_none test

mosesogwo committed:
update my_map test

mosesogwo committed:
update my_inject method and test

mosesogwo committed:
add rubocop corrections

mosesogwo committed:
update readme

JamesOkunlade reviewed (changes requested):
### **Good work with Tests in a first try!!! :slightly_smiling_face:** 

You have written nice tests for your enumerable methods and they pass well. What you'll need to do to make this perfect is to DRY your code.

You repeated test cases of arrays for every expected behavior you tested. You can do this once by using the `let` keyword to declare the array you'll be calling in each test rather than having to write them over and over again. You can check the [let block RSpec documentation ](https://github.com/rubocop-hq/rspec-style-guide#let-blocks) to get a better understanding and read the documentation to learn how to optimize your testing. 

Work on this and request another code review.

Good luck with it. :sunglasses:

**P.S**

This is a collaborative project. Why have you done it alone?

mosesogwo committed:
use let to hold array for tests

mosesogwo committed:
add rubocop correction

mosesogwo commented:
Thanks James.

I have effected the suggested correction.

My partner and I are doing this project solo because the original Enumerable Methods project was solo, so we have different codes for it.

mcrd25 commented on lib/enumerable.rb:
```diff
@@ -0,0 +1,116 @@
+  def my_each
+    each = []
+    if block_given?
+      self.each do |i|
```
Wh ydid you remove your original custom and replace it with this?

You are not supposed to use original enumerable method for your custom method.


I noticed before you had
```ruby
def my_each
    for i in self
      yield i
    end
  end
```

which is fine, you simply have to return self at the end and handle, if you so choose if a block is not given.


Because of this change I'd have to ask you to redo the tests for your actual custom method, not using each

mcrd25 commented on spec/enumerables_spec.rb:
```diff
@@ -0,0 +1,174 @@
+
+  describe '#my_each should return' do
+    it 'an array holding all elements of the given array' do
+      expect(arr.my_each).to eql([1, 2, 7, 8, 5])
```
**Tip: 💭** 😄 
This is fine for your custom method, however, if you look at the original method you'd notice that when a block is given it returns an enumerator hence a test like following would fail: 
`expect(arr.my_each).to eql(arreach)`

mcrd25 commented on spec/enumerables_spec.rb:
```diff
@@ -0,0 +1,174 @@
+
+    it 'an array of the yield for each element if given a block' do
+      result = arr.my_each { |x| x + 2 }
+      expect(result).to eql([3, 4, 9, 10, 7])
```
**Tip: 💭** 😄 
This is fine for your custom but actually, I strongly, strongly suggest you look at the original metho...

You would notice that it actually returns self when the block is given, i.e original array passed, so a test like following would fail:
```ruby
result = arr.my_each { |x| x + 2 }
result2 = arr.each { |x| x + 2 }
expect(result).to eql(result2)
```
What you can do to test something like this if you so choose to edit your original my_each to return self instead of the result of each. (Remember each is meant to really loop so the block should have all you need... if you needed to store result in an array you could implement that in the block

What yo ucan do to test for this is something like following
```ruby
describe 'my_each' do
    it 'calls the given block once for each element in self' do
      my_each_output = ''
      block = proc { |num| my_each_output += num.to_s }
      array.each(&block)
      each_output = my_each_output.dup
      my_each_output = ''
      array.my_each(&block)
      expect(my_each_output).to eq(each_output)
    end
  end
```

Instead of what it returns.


Hope this comment was valuable to you :)

mcrd25 commented on spec/enumerables_spec.rb:
```diff
@@ -0,0 +1,174 @@
+    end
+  end
+
+  describe '#my_each_with_index should return' do
```
Same comments  for my_each would apply to my_each_with_index

mcrd25 commented on spec/enumerables_spec.rb:
```diff
@@ -0,0 +1,174 @@
+    end
+
+    it 'the same array if no block is given' do
+      expect(arr.my_select).to eql([1, 2, 7, 8, 5])
```
*Suggestion*
Maybe you can try a test that runs against original select
SOmething like this would be nice 👍 
`expect(arr.my_select).to eql(arr.select)`

JamesOkunlade replied:
Yeah, this is cool. :grin:

mcrd25 reviewed (changes requested):
# Wow! I would have passed you but for one thing! Almost there! 😄 Almost 💯 

The one thing:
- **Your my_each method actually uses `each` which is not allowed. (see review comments for more info)**

Suggestions
- **You can also use `let` statements to store blocks. See the documentation of previous reviewer**
- **You should try and test your method against original as seen in review comments, you may even find something wrong with your custom method you did not know existed**


**Project looks great so far! The issue with the my_each is the problem! Even though your custom methods do not behave exactly like the original, that is fine! As your tests reflect how you expect your custom methods to behave! But please refactor your `my_each` to not include original `each` method. Thanks 😄**

***If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media.***

**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/)

mosesogwo committed:
remove each from my_each method

mosesogwo committed:
modify my_each with test

mosesogwo committed:
modify my_each, with_index, my_select and my_all again

mosesogwo committed:
Review all methods and test
Inject method not fully functional
three tests involving return of enumerator failing

mosesogwo committed:
rubocop layout

mosesogwo committed:
test against original methods

mosesogwo committed:
typo corrected

mosesogwo committed:
use take method to compare enumerators

mosesogwo committed:
add parenthesis to nested method calls

JamesOkunlade reviewed (approved):
### Good job. :+1:

You're good to go with this. Merge to master, close pull request and fill the project completion form.

Good luck with your future projects :+1:
